### ISSUE TITLE
cannot import name '_C' from 'groundingdino'

### ISSUE THREAD

**Issue by mvoodarla** [IDEA-Research/GroundingDINO#8]
I'm able to run it properly on a local GPU machine I've got but when I move this to the cloud on a docker image, I'm getting this issue. Have a feeling there's some system package I'm missing or something but not 100% sure.

**Comment by SlongLiu** [comment]
Thanks for your issue.

We have not tested the model on docker.  Can you provide more details of the outputs?

**Comment by mayoycy** [comment]
> move

try to copy your compiled codes into /opt/GLIP/ before starting training

**Comment by Innary** [comment]
@mvoodarla I think maybe it's because Rust is not installed.

**Comment by mvoodarla** [comment]
This is during inference, not training. It's likely some system package thing, I'll keep investigating.

**Comment by Innary** [comment]
> Hey I want to infer on my laptop i've totally installed repository however when i tried to infer with inference_on_a_image.py i took same error. How can i solve that

Perhaps it's caused by the lack of Rust-related environment installed on your system.

**Comment by Aleafy** [comment]
Grounded-Segment-Anything/GroundingDINO/groundingdino/models/GroundingDINO/ms_deform_attn.py:31: UserWarning: Failed to load custom C++ ops. Running on CPU mode Only!
  warnings.warn("Failed to load custom C++ ops. Running on CPU mode Only!")
final text_encoder_type: bert-base-uncased
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertModel: ['cls.predictions.transform.LayerNorm.bias', 'cls.predictions.decoder.weight', 'cls.seq_relationship.weight', 'cls.predictions.transform.dense.weight', 'cls.predictions.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.dense.bias', 'cls.seq_relationship.bias']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
_IncompatibleKeys(missing_keys=[], unexpected_keys=['label_enc.weight'])
/home/xxx/miniconda3/envs/sa/lib/python3.8/site-packages/transformers/modeling_utils.py:830: FutureWarning: The `device` argument is deprecated and will be removed in v5 of Transformers.
  warnings.warn(
/home/xxx/miniconda3/envs/sa/lib/python3.8/site-packages/torch/utils/checkpoint.py:25: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn("None of the inputs have requires_grad=True. Gradients will be None")
Traceback (most recent call last):
  File "grounded_sam_demo.py", line 145, in <module>
    boxes_filt, pred_phrases = get_grounding_output(
  File "grounded_sam_demo.py", line 57, in get_grounding_output
    outputs = model(image[None], captions=[caption])
  File "/home/xxx/miniconda3/envs/sa/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/xxx/perl5/Grounded-Segment-Anything/GroundingDINO/groundingdino/models/GroundingDINO/groundingdino.py", line 313, in forward
    hs, reference, hs_enc, ref_enc, init_box_proposal = self.transformer(
  File "/home/xxx/miniconda3/envs/sa/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/xxx/perl5/Grounded-Segment-Anything/GroundingDINO/groundingdino/models/GroundingDINO/transformer.py", line 258, in forward
    memory, memory_text = self.encoder(
  File "/home/xxx/miniconda3/envs/sa/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/xxx/perl5/Grounded-Segment-Anything/GroundingDINO/groundingdino/models/GroundingDINO/transformer.py", line 576, in forward
    output = checkpoint.checkpoint(
  File "/home/xxx/miniconda3/envs/sa/lib/python3.8/site-packages/torch/utils/checkpoint.py", line 177, in checkpoint
    return CheckpointFunction.apply(function, preserve, *args)
  File "/home/xxx/miniconda3/envs/sa/lib/python3.8/site-packages/torch/utils/checkpoint.py", line 75, in forward
    outputs = run_function(*args)
  File "/home/xxx/miniconda3/envs/sa/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/xxx/perl5/Grounded-Segment-Anything/GroundingDINO/groundingdino/models/GroundingDINO/transformer.py", line 785, in forward
    src2 = self.self_attn(
  File "/home/xxx/miniconda3/envs/sa/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/xxx/perl5/Grounded-Segment-Anything/GroundingDINO/groundingdino/models/GroundingDINO/ms_deform_attn.py", line 338, in forward
    output = MultiScaleDeformableAttnFunction.apply(
  File "/home/xxx/perl5/Grounded-Segment-Anything/GroundingDINO/groundingdino/models/GroundingDINO/ms_deform_attn.py", line 53, in forward
    output = _C.ms_deform_attn_forward(
NameError: name '_C' is not defined

**Comment by 1750585724** [comment]
I cannot install GroundingDINO. Can you help me?

**Comment by ByChelsea** [comment]
I installed Rust but this issue still exits... and I do not use docker :(
Can anyone help me? please...

**Comment by xu5zhao** [comment]
I encountered the same problem and solved the problem by looking at this [issue](https://github.com/IDEA-Research/Grounded-Segment-Anything/issues/15)!

The key is setting up CUDA_HOME correctly before installing repo~ here is an example：

```
export CUDA_HOME=/usr/local/cuda-11.6/
pip install -e .
```

If the GroundingDINO/groundingdino/_C.cpython-38-x86_64-linux-gnu.so file is generated, it indicates success.

If you still encounter problems, please refer to the issue under another project : https://github.com/IDEA-Research
/
[Grounded-Segment-Anything](https://github.com/IDEA-Research/Grounded-Segment-Anything) ,for a more detailed explanation.

**Comment by linyu92** [comment]
It seems unable to generate cpython files under windows platform while I type 'pip3 install -q -e .'

**Comment by linyu92** [comment]
> It seems unable to generate cpython files under windows platform while I type 'pip3 install -q -e .'

By the way, I have install Rust in windows platform

**Comment by darshats** [comment]
Hi,
We are facing the same error deploying Dino to azure ml. The build pipeline uses cpu machines. Hence entire setup.py is run on a cpu box.
First we had the failure of import torch in setup.py. This we got around by adding torch to conda.yml used by azure ml build pipeline. But this pulls in a cpu version of torch. The result is that at runtime when code is deployed to a gpu machine in azure, it complains about torch not being compiled for gpu.

Q:
1) is there a way we can simplify setup.py to not do dynamic compilation of the cpp files. Is there a solution where we can use precompiled _C.* files instead of creating on the fly? That would solve a whole lot of problems
2) the deformable attention seems to use interop to cpp. Is there any way to do this purely within pytorch and avoid #1.

**Comment by TemugeB** [comment]
A quick fix is to just use the pytorch implementation of deformable attention. 

Change this line:
https://github.com/IDEA-Research/GroundingDINO/blob/654f5e8bf97dce87da7e84e0d3feeb5bbad95388/groundingdino/models/GroundingDINO/ms_deform_attn.py#L330

to:
```python
        if not torch.cuda.is_available():
```

I don't know how much performance hit this will cause but since we are infering on single images, probably negligible.

**Comment by Kurinosuke118** [comment]
@mvoodarla How was this issue eventually resolved?

**Comment by Kurinosuke118** [comment]
@fkodom 
Does running python setup.py develop execute without modifying the requirements.txt of GroundingDINO?

**Comment by darshats** [comment]
I simply did as @TemugeB  suggested. Updated setup.py and removed the extension compilation and all torch references, also edited the ms_deform_attn.py to remove reference to the compiled _C function. It works, did not see any drop in inference quality, though my testing is limited.

**Comment by Kurinosuke118** [comment]
thank you!! I'll try it.

**Comment by lcolok** [comment]
First, let me provide the solution:
Add the TORCH_CUDA_ARCH_LIST environmental variable declaration in the Dockerfile, like this:
```bash
RUN git clone --depth=1 https://github.com/IDEA-Research/GroundingDINO.git
ENV TORCH_CUDA_ARCH_LIST="6.0;6.1;7.0;7.5;8.0;8.6+PTX;8.9;9.0"
RUN cd GroundingDINO/ && python -m pip install .
```
By doing this, the compiled GroundingDINO will not face the issue of "cannot import name '_C' from 'groundingdino'".

Now let me describe the process of finding this solution:https://github.com/IDEA-Research/GroundingDINO/issues/8#issuecomment-1485449302

Firstly, I want to thank @fkodom. Although his reply didn't solve my problem directly, it was very enlightening and made me realize that the issue might originate from the setup.py file.

By looking through the source code: https://github.com/IDEA-Research/GroundingDINO/blob/main/setup.py

I noticed a judgment statement in the script: `if CUDA_HOME is not None and (torch.cuda.is_available() or "TORCH_CUDA_ARCH_LIST" in os.environ):`

During the process of building the Docker image, you cannot use the GPU, because this process runs in the Docker daemon, which is isolated from your host machine and possible GPU hardware. Thus, when executing `torch.cuda.is_available()`, it will always return False. Relying solely on this condition, it can't perform the WITH_CUDA compilation, leading to the "cannot import name '_C' from 'groundingdino'" issue.

Fortunately, there's another condition: `"TORCH_CUDA_ARCH_LIST" in os.environ`. If we set the TORCH_CUDA_ARCH_LIST environment variable in the Dockerfile before executing setup.py, this problem can be solved. This environment variable specifies for which GPU architectures PyTorch should compile CUDA code. By setting this environment variable, we can clearly instruct PyTorch about the GPU architectures we want to compile CUDA code for.
For instance, we can add the following line in the Dockerfile:
```bash
ENV TORCH_CUDA_ARCH_LIST="6.0;6.1;7.0;7.5;8.0;8.6+PTX;8.9;9.0"
```
This would instruct PyTorch to compile CUDA code for the 6.0, 6.1, 7.0, 7.5, 8.0, and 8.6 architectures. These architectures cover a majority of common GPU device models, ensuring that our code will run on most GPU devices.

If you want to specify more precise definitions for your GPU model, you can refer to the official list from Nvidia: https://developer.nvidia.com/cuda-gpus

For example, all cards in the RTX 30 series are 8.6, while the RTX 40 series is 8.9.

If you are confused about the definition of GPU Compute Capability, you can get a better understanding by reading this article: https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/

**Comment by stevezkw1998** [comment]
Hi @fkodom I tried your Dockerfile
It errored on my side when running
![image](https://github.com/IDEA-Research/GroundingDINO/assets/53121066/e12b68e4-5e95-4e02-98f4-ae8abcf964f4)
ARG PY_VERSION="3.10"
```
E: Unable to locate package python-distutils
The command '/bin/sh -c add-apt-repository ppa:deadsnakes/ppa     && apt update     && apt install -y python$PY_VERSION-dev python$PY_VERSION-distutils     && update-alternatives --install /usr/bin/python python /usr/bin/python$PY_VERSION 10     && curl -sS https://bootstrap.pypa.io/get-pip.py | python     && apt clean     && rm -rf /var/lib/apt/lists/*' returned a non-zero code: 100
Docker build failed with error: Command 'docker build -t daat-backend-model:1.0.0 ..' returned non-zero exit status 100.
```
Do you have any idea how to solve it?

**Comment by eichi7** [comment]
> I simply did as @TemugeB suggested. Updated setup.py and removed the extension compilation and all torch references, also edited the ms_deform_attn.py to remove reference to the compiled _C function. It works, did not see any drop in inference quality, though my testing is limited.

How You did the setup.py updated and which extension You removed?

**Comment by yangchengxin** [comment]
> I encountered the same problem and solved the problem by looking at this [issue](https://github.com/IDEA-Research/Grounded-Segment-Anything/issues/15)!
> 
> The key is setting up CUDA_HOME correctly before installing repo~ here is an example：
> 
> ```
> export CUDA_HOME=/usr/local/cuda-11.6/
> pip install -e .
> ```
> 
> If the GroundingDINO/groundingdino/_C.cpython-38-x86_64-linux-gnu.so file is generated, it indicates success.
> 
> If you still encounter problems, please refer to the issue under another project : https://github.com/IDEA-Research / [Grounded-Segment-Anything](https://github.com/IDEA-Research/Grounded-Segment-Anything) ,for a more detailed explanation.

thanks for your sharing on this question, however, i have tried this solution and the problem is still existence. Would you please share the 'GroundingDINO/groundingdino/_C.cpython-38-x86_64-linux-gnu.so' file to me? this is my email:ycx971024@163.com

**Comment by Mehanik** [comment]
In my case the problem was  caused by ABI incompatibility. System GCC was used at building time, but python binary comes from conda. Installing gcc/gxx from conda solves the issue:
```
conda install gcc_linux-64
conda install gxx_linux-64
```

**Comment by apollonin** [comment]
For me, only the combination of the exported env and proper torch \ torchvision versions solved the problem!

```
export CUDA_HOME=/usr/local/cuda && export TORCH_CUDA_ARCH_LIST="6.0;6.1;7.0;7.5;8.0;8.6+PTX;8.9;9.0" && git clone https://github.com/IDEA-Research/GroundingDINO.git && cd GroundingDINO && pip install -e .
```

and requirements.txt:
```
torch==2.0.1
torchvision==0.15.2  # <--- did not work until I set the version to be 0.15.2
```

**Comment by teenaxta** [comment]
I was able to build docker with CUDA using the following image 
https://github.com/IDEA-Research/GroundingDINO/pull/307

**Comment by valentin-fontanger** [comment]
> A quick fix is to just use the pytorch implementation of deformable attention.
> 
> Change this line:
> 
> [GroundingDINO/groundingdino/models/GroundingDINO/ms_deform_attn.py](https://github.com/IDEA-Research/GroundingDINO/blob/654f5e8bf97dce87da7e84e0d3feeb5bbad95388/groundingdino/models/GroundingDINO/ms_deform_attn.py#L330)
> 
> Line 330 in [654f5e8](/IDEA-Research/GroundingDINO/commit/654f5e8bf97dce87da7e84e0d3feeb5bbad95388)
> 
>  if torch.cuda.is_available() and value.is_cuda: 
> to:
> 
>         if not torch.cuda.is_available():
> I don't know how much performance hit this will cause but since we are infering on single images, probably negligible.
> 
> Also make sure to send the model to cuda after you load it:
> 
> model = load_model("groundingdino/config/GroundingDINO_SwinT_OGC.py", "weights/groundingdino_swint_ogc.pth")
> model = model.to('cuda:0')

This should be pinged at the top somewhere for everyone to see so people don't have to waste days trying to make it work.
